actionmailbox/app/controllers/action_mailbox/ingresses/amazon/inbound_emails_controller.rb

```diff
@@ -0,0 +1,137 @@
+  #
+  # Returns:
+  #
+  # - <tt>204 No Content</tt> if a request is successfully processed
+  # - <tt>401 Unauthorized</tt> if a request does not contain a valid signature
+  # - <tt>404 Not Found</tt> if the Amazon ingress has not been configured
+  # - <tt>422 Unprocessable Entity</tt> if a request provides invalid parameters
+  #
+  # == Usage
+  #
+  # 1. Install the {aws-sdk-sns}[https://rubygems.org/gems/aws-sdk-sns] gem:
+  #
+  #        # Gemfile
+  #        gem "aws-sdk-sns", "~> 1.9", require: false
+  #
+  # 2. Tell Action Mailbox to accept emails from Amazon SES:
+  #
+  #        # config/environments/production.rb
+  #        config.action_mailbox.ingress = :amazon
+  #
+  # 3. Configure which SNS topics will be accepted:
+  #
+  #        config.action_mailbox.amazon.subscribed_topics = %w(
+  #          arn:aws:sns:eu-west-1:123456789001:example-topic-1
+  #          arn:aws:sns:us-east-1:123456789002:example-topic-2
+  #        )
+  #
+  # 4. {Configure SES}[https://docs.aws.amazon.com/ses/latest/DeveloperGuide/receiving-email-notifications.html]
+  #    to route emails through SNS.
+  #
+  #    Configure SNS to send emails to +/rails/action_mailbox/amazon/inbound_emails+.
+  #
+  #    If your application is found at <tt>https://example.com</tt> you would
+  #    specify the fully-qualified URL <tt>https://example.com/rails/action_mailbox/amazon/inbound_emails</tt>.
+  #
+  module Ingresses
+    module Amazon
+      class InboundEmailsController < ActionMailbox::BaseController
+        before_action :verify_authenticity
+        before_action :validate_topic
+        before_action :confirm_subscription
+
+        def create
+          head :bad_request unless mail.present?
+
+          ActionMailbox::InboundEmail.create_and_extract_message_id!(mail)
+          head :no_content
+        end
+
+        private
+          def verify_authenticity
+            head :bad_request unless notification.present?
+            head :unauthorized unless verified?
+          end
+
+          def confirm_subscription
+            return unless notification["Type"] == "SubscriptionConfirmation"
+            return head :ok if confirmation_response_code&.start_with?("2")
+
+            Rails.logger.error("SNS subscription confirmation request rejected.")
+            head :unprocessable_entity
+          end
+
+          def validate_topic
+            return if valid_topics&.include?(topic)
+
+            Rails.logger.warn("Ignoring unknown topic: #{topic}")
+            head :unauthorized
+          end
+
+          def confirmation_response_code
+            @confirmation_response_code ||= begin
+              Net::HTTP.get_response(URI(notification["SubscribeURL"])).code
+            end
+          end
+
+          def notification
+            @notification ||= JSON.parse(request.body.read)
+          rescue JSON::ParserError => e
+            Rails.logger.warn("Unable to parse SNS notification: #{e}")
+            nil
+          end
```

Comment:
Do we need to parse the request body manually, or can we rely on Rails parsing into `params` based on `Content-Type`?

Reply:
The reason for this is that the signature verification requires the exact structure provided in the post body. The `params` object contains extra parameters (`controller`, `action`) that invalidate the signature. I opted to use the same object in all places rather than mixing `params` and my parsed object.

Is there a way to extract just the post body JSON-extracted params from the `params` object ?

Reply:
`params.except(:controller, :action)` think that would be clearer too.

Reply:
@kaspth I considered this but it seems quite flaky to me. If any extra "special" parameters are added to Rails then this code will need to be updated. Since the SNS verifier depends on the exact input parameters included in the POST body, this [my implementation] seems the most correct way to do it, albeit with an extra `JSON.parse`.

I'm happy to be overruled !

Reply:
Don't recall us having added any extra default params in the last …10 years? My memory is a bit spotty beyond that, but don't think it's happened. I think the clarity of not sidestepping the default `params` parsing is overall better.

Is this something we could exercise in a test? Perhaps even with just a no-op action with `render json: params` and does `assert_empty params.except(:controller, :action)`?